delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/auth/OpenIdAuthService.scala

```diff
@@ -0,0 +1,70 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.auth
+
+import akka.http.javadsl.model.headers.HttpCredentials
+import akka.http.scaladsl.model.HttpMethods.POST
+import akka.http.scaladsl.model.headers.Authorization
+import akka.http.scaladsl.model.{HttpRequest, Uri}
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.kernel.Secret
+import ch.epfl.bluebrain.nexus.delta.kernel.effect.migration.MigrateEffectSyntax
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.sdk.auth.Credentials.ClientCredentials
+import ch.epfl.bluebrain.nexus.delta.sdk.error.AuthTokenError.{AuthTokenHttpError, AuthTokenNotFoundInResponse, ExpiryNotFoundInResponse}
+import ch.epfl.bluebrain.nexus.delta.sdk.http.HttpClient
+import ch.epfl.bluebrain.nexus.delta.sdk.realms.Realms
+import io.circe.Json
+import monix.bio.{IO, UIO}
+
+import java.time.{Duration, Instant}
+
+class OpenIdAuthService(httpClient: HttpClient, realms: Realms)(implicit clock: Clock[UIO])
+    extends MigrateEffectSyntax {
+
+  def auth(credentials: ClientCredentials): UIO[AccessTokenWithMetadata] = {
+    for {
+      realm                  <- realms.fetch(credentials.realm).toUIO
+      response               <- requestToken(realm.value.tokenEndpoint, credentials.user, credentials.password)
+      (token, validDuration) <- parseResponse(response)
+      now                    <- IOUtils.instant
+    } yield {
+      AccessTokenWithMetadata(token, now.plus(validDuration))
+    }
+  }
+
+  private def requestToken(tokenEndpoint: Uri, user: String, password: Secret[String]): UIO[Json] = {
+    httpClient
+      .toJson(
+        HttpRequest(
+          method = POST,
+          uri = tokenEndpoint,
+          headers = Authorization(HttpCredentials.createBasicHttpCredentials(user, password.value)) :: Nil,
+          entity = akka.http.scaladsl.model
+            .FormData(
+              Map(
+                "scope"      -> "openid",
+                "grant_type" -> "client_credentials"
+              )
+            )
+            .toEntity
+        )
+      )
+      .hideErrorsWith(AuthTokenHttpError)
+  }
+
+  private def parseResponse(json: Json): UIO[(String, Duration)] = {
+    for {
+      token  <- json.hcursor.get[String]("access_token") match {
+                  case Left(failure) => IO.terminate(AuthTokenNotFoundInResponse(failure))
+                  case Right(value)  => UIO.pure(value)
+                }
+      expiry <- json.hcursor.get[Long]("expires_in") match {
+                  case Left(failure) => IO.terminate(ExpiryNotFoundInResponse(failure))
+                  case Right(value)  => UIO.pure(value)
+                }
+    } yield {
+      (token, Duration.ofSeconds(expiry))
+    }
+  }
+}
+
+case class AccessTokenWithMetadata(token: String, expiresAt: Instant)
```

Comment:
My previous comment has been discarded it seems but the idea was:
* Rename to `TokenResponse`
* Match response fields names as they are part of the OpenId spec
* Use derivation with snake_case

Reply:
I'm not sure what the point of that would be? It's clearer with the interface how it is. We shouldn't need to care about the field names outside of integration with OpenId

Reply:
Let's forget about my previous comment.
I think I said it before then the comment got discarded and I forgot about it but we should use `ParsedToken` because we need the same validation than the one that is made by the identity module (with more checks checking that the issuer is present in the token and other things).
Otherwise we may end up with a token that is validated here, passed on to the storage and when the token gets back, it gets rejected by the identities endpoint